### PR TITLE
Add Snapping Behavior to DraggableScrollableSheet

### LINKED ISSUE #34111: Add drag length parameter to DraggableScrollableSheet
I am currently using heavy packages like [rubber](https://pub.dev/packages/rubber), [sliding_up_panel](https://pub.dev/packages/sliding_up_panel), [stopper](https://pub.dev/packages/stopper#-changelog-tab-) to implement what could be done w/ DraggableScrollableSheet if only it would snap open or closed upon drag release...

Only 1 new parameter would be needed: If null, the DraggableScrollableSheet would act as it does now; if set to a percentage of the parent, it would snap open or closed based on whether the threshold has been passed.

### PR DESCRIPTION
**Pull request by caseycrogers**
This PR adds two new optional arguments to `DraggableScrollableSheet` called `snap` and `snapSizes`. If `snap` is false (default value), the widget's behavior remains unchanged. If `snap` is true, the widget will snap to `snapSizes` when the user's finger leaves the screen during a drag. See for details: https://github.com/flutter/flutter/issues/34111#issuecomment-882141143.

This is achieved by replacing (if a snap should occur) the sheet's ballistic simulation in `_DraggableScrollableSheetExtent.goBallistic` with one that snaps to targets instead of moving forward until velocity is dissipated.

Fixes: #34111

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:153]: Made this a set because order isn't required and duplicate entries need not be supported. However, a list may be more intuitive for users. Should this remain a set or be made a list?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:198]: My IDE polluted the PR with all this garbage-I thought I was using the canonical Flutter fmt, if not how do I use it so I don't pollute commits? Or was this file originally committed with non-canonical formatting?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:300]: Broke this out so that it could be called directly from the snapping animation.

The logic of this function could be achieved in a less error prone fashion (ie it'd be harder to accidentally change the extent without sending a notification) with a listener on `_currentExtent`-AFAIK there are no circumstances where we want to set the `_currentExtent.value` without also emitting a scroll notification. Is that something that would be reasonable to add to this PR?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:152]: This is done for quality of life and because snapping behavior without including min and max as snap points seemed ill-defined/weird UX.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:528]: Waiting for the controller to stop makes snapping feel slow and clunky. Maybe we should snap when the velocity drops below a certain snapping threshold?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:550]: The curve and animation duration here were chosen fairly arbitrarily. Are this the right values and/or should these be user customizable?

**Review by caseycrogers — COMMENTED** [review]
Open questions/comments that I'd love to get feedback on from someone with more Flutter experience than me!

**Comment by caseycrogers** [comment]
@googlebot I signed it!

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:267]: This would be a natural place to use a set (order doesn't matter, duplicates don't make any conceptual sense). However, I hope to next tackle: https://github.com/flutter/flutter/issues/45009.

As a part of that PR I'd like to create a `Future<void> animateToSnapPoint(int index)` function where users can programmatically animate to a given snap point based on it's index in `snapPoints`. This functionality necessitates a list.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:154]: This would be a natural place to use a set (order doesn't matter, duplicates don't make any conceptual sense). However, I hope to next tackle: #45009.

As a part of that PR I'd like to create a Future<void> animateToSnapPoint(int index) function where users can programmatically animate to a given snap point based on it's index in snapPoints. This functionality necessitates a list.

Additionally, the current implementation assumes that you only ever want to snap to the closest target. We could let users specify snapping thresholds which would probably be an additionally list (or map?) and would required that `snapTargets` be both a list and sorted.

**Comment by caseycrogers** [comment]
@dnfield Hello Dan!

This PR closes out https://github.com/flutter/flutter/issues/34111. I noticed you've previously contributed to `DraggableScrollableSheet` and you were active in the above issue. This PR is still WIP (namely, behavior appears to work but there aren't any unit tests). Before I write my tests, I was wondering if you could give me a pre-review on the PR to let me know if I'm taking the right approach? Or direct me to the right reviewer?

File: examples/hello_world/lib/main.dart
Comment (dnfield) [inline examples/hello_world/lib/main.dart:5]: I assume this is your local test harness. I would recommend taking this and adding it as a unit test in the packages/flutter/test/widgets area. Please make sure to revert the changes here either way.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:154]: The behavior/expectations should be documented.

Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:154]: It also seems strange to me that an empty list automatically does anything. Do we always snap between min and max if values are otherwise specified?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:163]: Revert formatting chnages. This repo does not use dartfmt.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:269]: We should probably use the same type(s) here and in the widget, no?

**Review by dnfield — COMMENTED** [review]
/cc @goderbauer

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:154]: As is I haven't put any restrictions on `snapTargets`. If we later decide to add something like `animateToSnapPoint` or `snapThresholds`, we'd have to assert that `snapTargets` is an ordered list without duplicates. I'm wondering if it'd be better to add this assert in now to avoid a potential breaking change down the line?

We always snap between min and max. Under the hood I'm just adding min and max to the front and end of the user specified list, respectively. I agree that it's a little weird. Technically, you can specify min and max in `snapTargets`, it just won't have any effect. This mirrors the behavior of `Gradients.stops`.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:163]: Yeah sorry, I know that pollutes the PR and makes it harder to read. I intend to manually undo the formatting changes when the PR is done as I keep hitting the reformat hotkey out of habit.

File: examples/hello_world/lib/main.dart
Comment (caseycrogers) [reply examples/hello_world/lib/main.dart:5]: Yeah I just hijacked this as a testing playground-I'll revert the changes and create unit tests. I mostly wanted some feedback before building testing to make sure I'm on the right track.

**Comment by Alexays** [comment]
Just a comment about the behavior @caseycrogers, using `isScrollingNotifier.value` makes the UX weird, When the user removes the finger from the screen, the sheet should extend directly to the nearest snap point, instead of waiting until there is no more velocity (idle) and then extends.

**Comment by caseycrogers** [comment]
> Just a comment about the behavior @caseycrogers, using `isScrollingNotifier.value` makes the UX weird, When the user removes the finger from the screen, the sheet should extend directly to the nearest snap point, instead of waiting until there is no more velocity (idle) and then extends.

Yeah this is a good point. I've been pretty distracted lately but will get back to this in a bit. When I do I'll look for a way to snap immediately when the user's finger is released.

File: examples/hello_world/lib/main.dart
Comment (goderbauer) [inline examples/hello_world/lib/main.dart:13]: API-wise this looks strange. Why is the list empty?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:151]: nit: sentences in doc should follow proper grammar and start with a capital letter.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:152]: nit: spaces around "-"

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:151]: The first paragraph of a doc comment should be a single sentence summary.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:151]: What's the unit for snap points? Pixels? fractional values?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:154]: > It also seems strange to me that an empty list automatically does anything.

+1 to that.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:163]: To make the lives of your reviewers easier, please undo the unrelated formatting changes before the next round of reviews.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:163]: (I didn't fully review the rest because the extra diffs make it unnecessarily painful).

**Review by goderbauer — COMMENTED** [review]
Please also add tests to this PR.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:163]: Yeah sorry-I just wanted really quick directional feedback (eg right approach wrong approach) before investing a lot more time into the PR (polish, tests etc.). It's not ready for full review, I'll take it out of draft once it is ready for review.

Thanks!

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:154]: If snapping is enabled, it's not very coherent to NOT treat min and max as snapping targets-what would it mean to have parts of the draggable range snap and other parts not?

Right now I'm using a non-null `snapTargets` to indicate that snapping is desired. How about I:
- Introduce another argument `Bool snap` that enables or disables snapping
- Retain current behavior where `snapTargets` always implicitly includes min and max whether or not the user specifies them (consistent with https://api.flutter.dev/flutter/painting/Gradient/stops.html)
- Make `snapTargets` a non-nullable list with the empty list as the default value

This way the user never needs to provide an empty list to trigger snapping behavior.

**Comment by caseycrogers** [comment]
Waiting on https://github.com/flutter/flutter/pull/86614.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:170]: Implemented the above solution.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:496]: There was some weird behavior here where this variable was named and initialized as if it were a delta, but was being used in `_tick` as an absolute position. Initializing it to the wrong value (0 as opposed to the current extent in pixels) wasn't causing any errant behavior because the error was being subtracted back out on L498, but I've fixed this anyway because this could introduce a confusing bug if L498 ever gets refactored.

**Comment by caseycrogers** [comment]
@dnfield hey Dan! Thanks for all your time so far. This PR is ready for review-could take a look at it whenever you have the chance?

**Comment by Alexays** [comment]
@caseycrogers Any ideas how to implement updateExtent with animation?
Thanks for your changes!

**Comment by caseycrogers** [comment]
> @caseycrogers Any ideas how to implement updateExtent with animation?
> Thanks for your changes!

Hey @Alexays, I'm not quite sure what you mean, but I could probably help if you ask me in the flutter discord channel (just @, same username as GitHub) in the general channel.

**Comment by caseycrogers** [comment]
@dnfield Hey Dan! Bumping this in case you missed it last time. CC @goderbauer in case you have the time to review my PR.

Also, let me know if I should ask elsewhere for review.

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 53.
View them at https://flutter-gold.skia.org/cl/github/84394

**Comment by dnfield** [comment]
Posted in the issue but is https://pub.dev/packages/snapping_sheet a better solution in this case?

**Comment by caseycrogers** [comment]
> Posted in the issue but is https://pub.dev/packages/snapping_sheet a better solution in this case?

Responded [in the issue](https://github.com/flutter/flutter/issues/34111#issuecomment-899783961)!

**Comment by caseycrogers** [comment]
@dnfield let me know if you want any more input from me before reviewing!

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: What do you think about creating a specific object for this instead? 

Asking because when I read the param list I found myself wondering why there are two properties for snapping and what it means to set one but not the other. There are also some rules about what goes into the list of snap points and what kind of numbers they have to be. 

I suspect that if we created a class for this it would be a bit more self-explanatory and harder to misuse. WDYT?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:365]: Dart generally says to avoid `forEach` unless you're applying a function to it directly. This should be rewritten as a regular for loop.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:399]: Is this right? Dependencies can change after a user has interacted with the sheet.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:548]: Flutter style says not to use `=>` unless it's a short, one line function. There are some additional formatting issues below (it's ok to exceed 80 chars if it's more readable).

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:496]: Nah this is fine.

File: packages/flutter/test/widgets/draggable_scrollable_sheet_test.dart
Comment (dnfield) [inline packages/flutter/test/widgets/draggable_scrollable_sheet_test.dart:11]: Please avoid introducing new epsilons other than the one in `constants.dart` from foundation.

**Review by dnfield — COMMENTED** [review]
Looking closer!

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 54.
View them at https://flutter-gold.skia.org/cl/github/84394

**Review by caseycrogers — COMMENTED** [review]
@dnfield Resolved formatting comments. I had two questions (see open conversations), once I get more feedback I'll push a commit and resolve them.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:402]: Ahh ok, I see now. The way the diff was rendered by default was confusing me.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: I think it's confusing to have a bool argument and a field - it's either there or not, so null or not null seems to make sense.

That said, if you wanted an `enabled` property on `SnapConfiguration` I wouldn't really complain either.

In Dart, newing up objects is pretty cheap, and with named properties I think it should be fairly idiomatic. I'm suggesting switching to a class here :)

Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: @dnfield bump

Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: @dnfield bump?

Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: Asserting things about what's in the list means this can't be const anymore, which may end up being breaking. 

I think if we just use a `SplayTreeSet` it should work - which is similar to what we end up doing in gradient.dart to make sure the stops and colors are sorted and unique anyway.

Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: (We could still do the asserts at time of use instead of time of construction of course)

Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: Do you mean use `SplayTreeSet<double>` as our input type for `snapPoints`?

We can't enforce that the input set is using least-to-greatest ordering can we? We could Take in an arbitrarily ordered list and then convert it to a LTG ordered `SplayTreeSet` internally. I marginally prefer a list that is asserted to be LTG ordered, but I'm fine with either.

Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: `SplayTreeSet<double>` will automatically be ordered least to greatest with duplicates removed.

Keeping it as a list and documenting the expectation is fine - it could be converted to a SplayTreeSet later. Having the asserts is fine too, but they can't be on the ctor without dropping const, so they'd have to be on `build` or `initState` or something like that.

**Comment by b3nni97** [comment]
Hi I have tested your PR and would prefer to have a control how the snapping animation runs. (setting curves etc.)

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:178]: Awesome! If we're okay with a flat list and asserts, then I think we're ready to move forward as is?

`_impliedSnapSizes` asserts that the list is properly ordered and between min and max before it is used:
https://github.com/flutter/flutter/pull/84394/files#diff-8a2b7e9da235db9760f0046cbbceddc5e41c78c9285a315141c401f715bd269fR364

and there's a test that ensures the asserts are thrown on invalid input:
https://github.com/flutter/flutter/pull/84394/files#diff-6d377811be6bfec7ab69f334aca4ddbaa49baf8a6639103e86afa248919d5e75R95

Let me know if there are any other blocking issues you see!

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:353]: @dnfield It's just occurred to me that this code in prod and in my branch is a bit broken:
`DraggableScrollableSheet` does not update the extent when the widget is rebuilt with new values. This means that the extent will retain potentially invalid values for min, max, snapSizes, etc. if the user rebuilds the sheet with new parameters.

We need to override `DidUpdateWidget` to rebuild the extent when necessary. This feels fairly non-trivial and is largely about fixing existing behavior in prod-are you okay with me leaving this fix to a follow up PR to go in after this one merges? Or should I spin up a new PR now to fix it and rebase this one on top of it? Seems like a fairly severe bug that may justify the latter.

Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:353]: Here's a test that can be added to `draggable_scrollable_sheet_test.dart` to demonstrate the issue:
```dart
    testWidgets('A changed snapSizes list is propagated', (WidgetTester tester) async {
      const Key stackKey = ValueKey<String>('stack');
      const Key containerKey = ValueKey<String>('container');
      await tester.pumpWidget(_boilerplate(null,
        snap: true,
        stackKey: stackKey,
        containerKey: containerKey,
        snapSizes: <double>[.5],
      ));
      await tester.pumpAndSettle();
      final double screenHeight = tester.getSize(find.byKey(stackKey)).height;

      await tester.drag(find.text('Item 1'), Offset(0, -.4 * screenHeight));
      await tester.pumpAndSettle();
      expect(
        tester.getSize(find.byKey(containerKey)).height / screenHeight,
        closeTo(1.0, precisionErrorTolerance),
      );

      // Pump the same widget but with a new snapSizes, triggering a rebuild.
      await tester.pumpWidget(_boilerplate(null,
        snap: true,
        stackKey: stackKey,
        containerKey: containerKey,
        maxChildSize: .8,
        snapSizes: <double>[.5],
      ));
      await tester.pumpAndSettle();

      // The max child size has been reduced, we should be rebuilt at the new
      // max of .8.
      expect(
        tester.getSize(find.byKey(containerKey)).height / screenHeight,
        closeTo(.8, precisionErrorTolerance),
      );
    }, variant: TargetPlatformVariant.all());
```

Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:353]: Doing this in a new PR is fine. https://github.com/flutter/flutter/pull/69724/files probbly would fix this, and relanding that without the bit about caching the child is probably a good idea.

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:368]: nit: here and below, probably good to dump the whole list and point to the invalid one rather than just printing the invalid one, which might occur multiple times in a badly formed list.

**Review by dnfield — APPROVED** [review]
LGTM with nitpick.

Thanks for sticking with this.

If you plan to support supplying custom curves for the animation, maybe we can do that in a separate PR?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart]: Fix'd!

```
Invalid snapSize '0.4' at index 3 of:
  [0.3, 0.4, 0.5, >>> 0.4 <<<, 0.6, 0.7]
Snap sizes must be in ascending order.
```

However, I'm getting the canned Flutter message "Either the assertion indicators an error in the framework itself, or we should..." message. It looks like it's just naively checking if the top two frames of the stack are Flutter and not user code? Do you know how to get rid of this?

**Comment by caseycrogers** [comment]
> LGTM with nitpick.
> 
> Thanks for sticking with this.
> 
> If you plan to support supplying custom curves for the animation, maybe we can do that in a separate PR?

Happy to do that in a separate PR! I fixed the nit, but there's still one minor issue (responded in comment thread).

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart]: Feel free to file a bug for that - that code has a history of breaking :\

Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart]: File a bug for moving the asserts into the constructor, or file a bug for the "Either the assertion..." message firing?

Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart]: For the `Either the assertion` message :)

**Comment by caseycrogers** [comment]
@dnfield How do we get this merged? It looks like the tests are permanently pending?

**Comment by dnfield** [comment]
Hm. Can you push up an empty commit to kick CI? Something seems to have gotten hung up.

**Comment by caseycrogers** [comment]
Things appear very busted :/. I tried `git fetch upstream` and `git merge master`, neither fetched any new commits. should I post in hackers discord?

**Comment by dnfield** [comment]
Looks like ci is running now

**Comment by Piinks** [comment]
It looks like the git issue here may have broken the flutter-gold check. I will watch this and merge manually once the Google testing shard passes, hoping this will unblock the flutter-gold check.

**Comment by Piinks** [comment]
The google check also appears broken here, I am going to merge this manually. I confirmed the google testing shard passed.

**Comment by MalteMagnussen** [comment]
Thanks a lot for this! 
Great job @caseycrogers

**Pull request merged**
